File: frontend/src/lib/vendorType.js
export const VENDOR_TYPES = {
  SERVICE: 'service_vendor',
  PRODUCT: 'product_vendor',
};

const PRODUCT_CATEGORY_IDS = new Set(['cat-grocery', 'grocery']);
const PRODUCT_VALUES = new Set(['product_vendor', 'product', 'grocery']);
const SERVICE_VALUES = new Set(['service_vendor', 'service']);

export function resolveVendorType(vendor) {
  if (!vendor) return VENDOR_TYPES.SERVICE;
  const explicit = String(vendor.vendor_type || '').toLowerCase();
  if (PRODUCT_CATEGORY_IDS.has(vendor.category_id)) {
    if (explicit && !PRODUCT_VALUES.has(explicit)) {
      console.warn('Vendor type mismatch for grocery category; forcing PRODUCT_VENDOR', vendor);
    }
    return VENDOR_TYPES.PRODUCT;
  }
  if (vendor.category_id) {
    if (PRODUCT_VALUES.has(explicit)) {
      console.warn('Vendor type mismatch for service category; forcing SERVICE_VENDOR', vendor);
    }
    return VENDOR_TYPES.SERVICE;
  }
  if (PRODUCT_VALUES.has(explicit)) {
    return VENDOR_TYPES.PRODUCT;
  }
  if (SERVICE_VALUES.has(explicit)) {
    return VENDOR_TYPES.SERVICE;
  }
  if (!vendor.category_id) {
    console.warn('Vendor type unclear; defaulting to SERVICE_VENDOR', vendor);
  }
  return VENDOR_TYPES.SERVICE;
}

export function isProductVendor(vendor) {
  return resolveVendorType(vendor) === VENDOR_TYPES.PRODUCT;
}

File: backend/services/vendor_type.py
from __future__ import annotations

import logging
from typing import Optional

from models import VendorType

logger = logging.getLogger(__name__)

PRODUCT_CATEGORY_IDS = {"cat-grocery", "grocery"}
PRODUCT_VALUES = {"product_vendor", "product", "grocery"}
SERVICE_VALUES = {"service_vendor", "service"}


def resolve_vendor_type(category_id: Optional[str] = None, explicit: Optional[str] = None) -> VendorType:
    explicit_value = str(explicit).lower().strip() if explicit is not None else ""

    if category_id in PRODUCT_CATEGORY_IDS:
        if explicit_value and explicit_value not in PRODUCT_VALUES:
            logger.warning(
                "Vendor type mismatch for grocery category_id=%s explicit=%s; using PRODUCT_VENDOR",
                category_id,
                explicit,
            )
        return VendorType.PRODUCT_VENDOR

    if category_id:
        if explicit_value in PRODUCT_VALUES:
            logger.warning(
                "Vendor type mismatch for service category_id=%s explicit=%s; using SERVICE_VENDOR",
                category_id,
                explicit,
            )
        return VendorType.SERVICE_VENDOR

    if explicit_value in PRODUCT_VALUES:
        return VendorType.PRODUCT_VENDOR
    if explicit_value in SERVICE_VALUES:
        return VendorType.SERVICE_VENDOR

    logger.warning("Vendor type could not be resolved; defaulting to SERVICE_VENDOR")
    return VendorType.SERVICE_VENDOR


def ensure_vendor_type(vendor_doc: dict) -> VendorType:
    existing = vendor_doc.get("vendor_type") if vendor_doc else None
    resolved = resolve_vendor_type(vendor_doc.get("category_id") if vendor_doc else None, existing)
    vendor_doc["vendor_type"] = resolved.value
    return resolved
